core/src/main/scala/org/apache/spark/sql/CachedDataFrame.scala

```diff
@@ -128,136 +127,142 @@ class CachedDataFrame(session: SparkSession, queryExecution: QueryExecution,
-    lastExecRDD = cachedRDD
-  }
-  private lazy val rdd_partitions_ = {
-    val _f = classOf[RDD[_]].getDeclaredField("org$apache$spark$rdd$RDD$$partitions_")
-    _f.setAccessible(true)
-    UnsafeHolder.getUnsafe.objectFieldOffset(_f)
+  private[sql] def duplicate(): CachedDataFrame = {
+    val cdf = new CachedDataFrame(snappySession, queryExecution, queryExecutionString,
+      queryPlanInfo, currentQueryExecutionString, currentQueryPlanInfo, cachedRDD,
+      shuffleDependencies, encoder, rddId, noSideEffects, queryHints, -1L, -1L, -1L)
+    cdf.log_ = log_
+    cdf.levelFlags = levelFlags
+    cdf._boundEnc = _boundEnc
+    cdf._rowConverter = _rowConverter
+    cdf.paramLiterals = paramLiterals
+    cdf
   }
 
-  @tailrec
-  private def clearPartitions(rdd: Seq[RDD[_]]): Unit = {
-    val children = rdd.flatMap(r => if (r != null) {
-      r.dependencies.map {
-        case d: NarrowDependency[_] => d.rdd
-        case s: ShuffleDependency[_, _, _] => s.rdd
+  private def reset(): Unit = clearPartitions(cachedRDD :: Nil)
+
+  private def applyCurrentLiterals(): Unit = {
+    if (paramLiterals.length > 0 && (paramLiterals ne currentLiterals)) {
+      for (pos <- paramLiterals.indices) {
+        paramLiterals(pos).updateValue(currentLiterals(pos).value)
       }
-    } else None)
+    }
+  }
 
-    rdd.foreach {
-      case null =>
-      case r: RDD[_] =>
+  @tailrec
+  private def clearPartitions(rdds: Seq[RDD[_]]): Unit = {
+    val children = rdds.flatMap {
+      case null => Nil
+      case r =>
         // f.set(r, null)
-        UnsafeHolder.getUnsafe.putObject(r, rdd_partitions_, null)
-    }
-    if (children.isEmpty) {
-      return
+        UnsafeHolder.getUnsafe.putObject(r, CachedDataFrame.rdd_partitions_, null)
+        r.dependencies.map(_.rdd)
     }
-    clearPartitions(children)
+    if (children.nonEmpty) clearPartitions(children)
   }
 
   private def setPoolForExecution(): Unit = {
-    var pool = sparkSession.asInstanceOf[SnappySession].
-      sessionState.conf.activeSchedulerPool
-
+    var pool = snappySession.sessionState.conf.activeSchedulerPool
     // Check if it is pruned query, execute it automatically on the low latency pool
-    if (isLowLatencyQuery && shuffleDependencies.length == 0 && pool == "default") {
-      if (sparkSession.sparkContext.getAllPools.exists(_.name == "lowlatency")) {
-        pool = "lowlatency"
+    if (isLowLatencyQuery && shuffleDependencies.isEmpty && pool == "default") {
+      if (snappySession.sparkContext.getPoolForName(Constant.LOW_LATENCY_POOL).isDefined) {
+        pool = Constant.LOW_LATENCY_POOL
       }
     }
-    sparkSession.sparkContext.setLocalProperty("spark.scheduler.pool", pool)
+    snappySession.sparkContext.setLocalProperty("spark.scheduler.pool", pool)
+  }
+
+  private def prepareForCollect(): Unit = {
+    if (isCached) {
+      reset()
+      applyCurrentLiterals()
+    }
+    setPoolForExecution()
```

Comment:
Currently I don't think low-latency check can change but in future if we add more partitioning schemes it can. Secondly this will also honor any current "spark.scheduler.pool" setting on the session.